man/org.freedesktop.systemd1.xml

```diff
@@ -1130,7 +1130,7 @@ node /org/freedesktop/systemd1 {
 
       <para><function>StopUnit()</function> is similar to <function>StartUnit()</function> but stops the
       specified unit rather than starting it. Note that the <literal>isolate</literal> mode is invalid for this
-      method.</para>
+      method. Mode <literal>replace-unload</literal> will cause the unit to be immediately unloaded after stopping.</para>
```

Comment:
this appears unnecessary. systemd should GC everything not referenced anymore instantly anyway. This behaviour should hence not be an option, but the implicit non-configurable default.

A unit only remains loaded if the GC determines that it is referenced by some other unit or if it is running or failed. "systemctl show -p ReferencedBy" should tell you which other unit references your unit and thus keeps it in memory. You can tweak the "failed" thing with `CollectMode=inactive-or-failed` btw. 

Hence, instead of this commit, are you sure that CollectMode= doesn't solve your issue sufficiently? And if not, can you try to track down what other unit keeps it pinned via systemctl show?

Reply:
Will double check again - I'm pretty sure it was an After/Before statement with another portable service that caused it to be kept around.

Reply:
Yes, an "After" is sufficient. The testsuite-48 I added is enough to reproduce this (it uses Wants currently, can be just swapped to After). "ReferencedBy" returns nothing.

```
root@localhost:/tmp# cat > /run/systemd/system/testservice-48.target <<EOF
> [Unit]
> After=testservice-48.service
> EOF
root@localhost:/tmp# systemctl daemon-reload
root@localhost:/tmp# systemctl start testservice-48.target
root@localhost:/tmp# cat > /run/systemd/system/testservice-48.service <<EOF
> [Unit]
> CollectMode=inactive-or-failed
> 
> [Service]
> ExecStart=/bin/sleep 100000000
> Type=exec
> EOF
root@localhost:/tmp# systemctl start testservice-48.service
root@localhost:/tmp# systemctl is-active testservice-48.service
active
root@localhost:/tmp# systemctl stop testservice-48.service
root@localhost:/tmp# rm -f /run/systemd/system/testservice-48.service
root@localhost:/tmp# systemctl show -p ReferencedBy testservice-48.service
root@localhost:/tmp# systemctl status testservice-48.service
● testservice-48.service
     Loaded: loaded (/run/systemd/system/testservice-48.service; bad; vendor preset: enabled)
   Active: inactive (dead) (thawing) since Tue 2020-05-19 10:26:06 UTC; 1min 3s ago
    Process: 710 ExecStart=/bin/sleep 100000000 (code=killed, signal=TERM)
   Main PID: 710 (code=killed, signal=TERM)
        CPU: 88ms

May 19 10:25:50 localhost systemd[1]: Starting testservice-48.service...
May 19 10:25:51 localhost systemd[1]: Started testservice-48.service.
May 19 10:26:06 localhost systemd[1]: Stopping testservice-48.service...
May 19 10:26:06 localhost systemd[1]: testservice-48.service: Succeeded.
May 19 10:26:06 localhost systemd[1]: Stopped testservice-48.service.
Warning: The unit file, source configuration file or drop-ins of testservice-48.service changed on disk. Run 'systemctl daemon-reload' to reload units.
root@localhost:/tmp# systemctl stop --job-mode replace-unload testservice-48.service
root@localhost:/tmp# systemctl status testservice-48.service
Unit testservice-48.service could not be found.
```

Reply:
consider reversing the dep.

References= and ReferencedBy= are internal deps that are synthesized whenever you configure any other dep (for example After= or Wants) between a unit. References= from the unit that declares that dep to the other. The GC just cares for References= and ReferencedBy=  to figure out what it can unload.

hence, if you have a unit A that has After=B then this also implies that A References=B. And this means that B is not GC'ed until A is GC'ed first. Since After= and Before= are inverse of each other you can thus control which way you want to do things: for the runtime A with After=B is identical to B with Before=A. However for the GC logic due to these implicit References= deps the former means A needs to be GC'ed before B will be while the latter means the opposite.

Reply:
That can work in some cases - but consider the example of a portable service that one wants to start automatically on boot - eg, portablectl attach --enable with the unit file having the usual WantedBy=some.target. In that case, the target will always have a reference to it, so on detach (eg: application upgrade to a new version) it will not be garbage collected.

Reply:
Eg, with a slight variation from testsuite-49:

```
root@localhost:~# portablectl attach --copy=symlink --now --no-reload --enable /var/img testservice-49
(Matching unit files with prefixes 'testservice-49'.)
Created directory /etc/systemd/system.attached.
Created directory /etc/systemd/system.attached/testservice-49.target.d.
Written /etc/systemd/system.attached/testservice-49.target.d/20-portable.conf.
Created symlink /etc/systemd/system.attached/testservice-49.target → /var/img/usr/lib/systemd/system/testservice-49.target.
Created symlink /etc/portables/img → /var/img.
Created symlink /etc/systemd/system/multi-user.target.wants/testservice-49.target → /var/img/usr/lib/systemd/system/testservice-49.target.
Created symlink /etc/systemd/system/testservice-49.target → /var/img/usr/lib/systemd/system/testservice-49.target.
Queued /org/freedesktop/systemd1/job/2112 to start portable service testservice-49.target.
root@localhost:~# systemctl status testservice-49.target
● testservice-49.target - I am portable!
     Loaded: loaded (/var/img/usr/lib/systemd/system/testservice-49.target; enabled; vendor preset: enabled)
    Drop-In: /etc/systemd/system.attached/testservice-49.target.d
             └─20-portable.conf
     Active: active since Tue 2020-05-19 13:23:55 UTC; 14s ago

May 19 13:23:55 localhost systemd[1]: Reached target I am portable!.
root@localhost:~# portablectl detach --enable --now --no-reload /var/img testservice-49
(Matching unit files with prefixes 'testservice-49'.)
Queued /org/freedesktop/systemd1/job/2390 to stop portable service testservice-49.target.
Removed /etc/systemd/system/multi-user.target.wants/testservice-49.target.
Removed /etc/systemd/system/testservice-49.target.
Removed /etc/systemd/system.attached/testservice-49.target.
Removed /etc/systemd/system.attached/testservice-49.target.d/20-portable.conf.
Removed /etc/systemd/system.attached/testservice-49.target.d.
Removed /etc/portables/img.
Removed /etc/systemd/system.attached.
root@localhost:~# systemctl status testservice-49.target
● testservice-49.target - I am portable!
     Loaded: loaded (/var/img/usr/lib/systemd/system/testservice-49.target; bad; vendor preset: enabled)
   Active: inactive (dead)

May 19 13:23:55 localhost systemd[1]: Reached target I am portable!.
```

Reply:
@poettering gentle ping :-) What do you think about the above?
Thanks!

Reply:
@poettering one more ping, first part was merged independently. Would like to bring this one forward too. Thanks!